Clean up run.sh

hachibu opened the pull request:
Clean up the following in run.sh:
* Fix shellcheck warnings.
* Switch to #!/usr/bin/env bash to use the current environment's bash.
* Parse command line flags.
* Add help message.
* Use for loop to dry code up when checking dependencies.
* Send pushd and popd stdout and stderr to /dev/null to declutter script output.
* Check multirust list-toolchains for existing install before overriding.
* Store multirust version in a variable.

Minor:
* Reword a few sentences.

<img width="745" alt="screen shot 2015-08-25 at 5 16 32 pm" src="https://cloud.githubusercontent.com/assets/10385848/9482826/11d5def6-4b4d-11e5-9288-698e26ea5762.png">

<img width="754" alt="screen shot 2015-08-25 at 3 51 04 pm" src="https://cloud.githubusercontent.com/assets/10385848/9481668/24637de6-4b41-11e5-8e47-96d7d2a8a8c8.png">

joshuafcole commented on run.sh:
```diff
@@ -1,36 +1,74 @@
-  tsc
+# Try using the TypeScript compiler (tsc) to compile UI.
+pushd . &> /dev/null
+  printf "* Compiling editor..."
```
Out of curiosity, why printf instead of echo here?

hachibu replied:
:+1:

joshuafcole commented on run.sh:
```diff
@@ -1,36 +1,74 @@
-# If we aren't restarting, open the editor in the user's preferred browser
-if [[ "x$1" != "x--restart" ]]; then
-  echo "* Opening $waitUrl"
+# If openBrowserFlag is set open the editor in the user's preferred browser.
```
I think we probably want to do this by default. Better to have two editor windows than zero as a first time user experience. Thoughts?

hachibu replied:
@joshuafcole I agree with you. It is a better user experience for the editor to just pop up.

Maybe we can pick a better flag name that's more descriptive?

joshuafcole replied:
Yeah, I agree. Perhaps `--headless`? Or the more verbose `--noBrowser`?

hachibu replied:
@joshuafcole I really like --noBrowser. What do you think about --no-browser to make it more Unix-y?

joshuafcole replied:
That's fine. I should update the test.sh script to match. I left it lowerCamel to make the casperjs interop simpler, but I can paper over that on the JS side I guess.

hachibu replied:
:+1:

joshuafcole commented on run.sh:
```diff
@@ -38,18 +76,24 @@ if [[ "x$1" != "x--restart" ]]; then
-  # Compile runtime server
-  echo "* Compiling server... (This takes a while)"
+  # Ensure rustc is updated.
+  multirust list-toolchains | grep "$rustVersion" &> /dev/null
```
@jamii Do we still need to check that this is the right version of rust currently?

joshuafcole replied:
I confirmed that we need to override irrespective of whether it's installed or not. Since multiple rust versions can be installed in multi-rust, we need to tell it to use the appropriate version. I do like the clearer diagnostic output though. Also, for consistency, in the case that rust does need updated it would be nice to use printf and show completion like you are elsewhere.

hachibu replied:
@joshuafcole Cool. Can you elaborate more on the case that could be more consistent?

joshuafcole replied:
If we detect that rust doesn't have `$rustVersion` installed, we know `multirust override "$rustVersion"` will trigger an installation. Since we echo "* Updating rust..." when this happens, and since multirust is blocking, it would be nice to printf that message and echo "done." when it completes the way you have for the other build phases.

hachibu replied:
Ok, I understand now. Thank you. I will patch this up.

joshuafcole replied:
This is the last issue, then we're good to go.

hachibu replied:
Awesome. My commit is going through CI.

joshuafcole commented:
Everything else looks great by the way. Once those two things get updated I'll be happy to merge this in. Thanks for your help.

joshuafcole commented on run.sh:
```diff
@@ -1,36 +1,74 @@
+deps="tsc multirust"
+for dep in $deps; do
+  if ! which "$dep" &> /dev/null; then
+    printf "  x Please install %s: " "$dep"
```
Perhaps we should leave the X on this line but print the fix on the next line? It looks a bit cluttered on a single line, and some people (though I can't fathom why) still use 80 character terminal widths.

![screen shot 2015-08-25 at 4 56 27 pm](https://cloud.githubusercontent.com/assets/313870/9482553/56448ce8-4b4a-11e5-9e45-2afbe36f20e6.png)

hachibu replied:
:+1: I agree. I'll place it on the next line.

hachibu replied:
@joshuafcole what do you think about this? 

<img width="547" alt="screen shot 2015-08-25 at 5 13 12 pm" src="https://cloud.githubusercontent.com/assets/10385848/9482780/9d530f90-4b4c-11e5-855d-1544fb29e974.png">

hachibu replied:
I accidentally left out the sudo. But I'll put it back in.

joshuafcole replied:
Yep, that works.

joshuafcole commented on run.sh:
```diff
@@ -1,36 +1,74 @@
+  printf "* Compiling editor..."
+  cd ui
+  tscError=$(tsc)
   if [ $? -ne 0 ]; then
```
`set -e` means we never get into this branch, and since you capture the output of tsc instead of letting it write to screen, we lose the failure output. I'm not sure what the right solution here is though.

hachibu replied:
@joshuafcole I just ran into this! I think I was a little over-zealous adding set -e. I'm going to remove it.

hachibu commented:
@joshuafcole when you feel like this is good to go I can squash this into a single commit.

hachibu commented:
@joshuafcole I just pushed the multirust override fix. When you get a chance take another look over everything and then I can squash.

joshuafcole commented:
Looks great, squash when you get the chance and we'll get it merged. Thanks again for the contribution.

ibdknox commented:
@hachibu Hey Ray, I just wanted to say thanks for all the work here! It is greatly appreciated :)

hachibu commented:
@joshuafcole @ibdknox thanks guys! i'm really happy to contribute.

I just updated this branch, squashed my commits and CI has passed so it should be ready to merge.
